### ISSUE TITLE
playback rate / sound pitch

### ISSUE THREAD

**Issue by pitrackster** [katspaugh/wavesurfer.js#149]
Changing playback rate is changing sound pitch.

As this behavior could be needed, sometimes we could need to preserve the original pitch (even with artifacts).

**Comment by yoyomo** [comment]
It seems everyone has ignored the OP's problem. 

> Changing playback rate is changing sound pitch.

we don't want that. HTML5 <audio> tag can neatly do this without changing pitch.

**Comment by sprycoder** [comment]
First of all Thanks to team for creating such a useful component for audio. 

I am using it for my project and its really easy to use. But I'm disappointed by the change in sound pitch after increasing playbackRate. I tried with waveSufer.setPlaybackRate, waveSurfer.backend.setPlaybackRate and other few options.

Is it still an open bug or I have to do it different way?

**Comment by jeremylach** [comment]
I'd also be interested in this feature.

**Comment by katspaugh** [comment]
I'll need to do some research on this.

**Comment by jsphweid** [comment]
@katspaugh I've been looking into this and wanted to present this to you.

Someone [created a javascript version of the famous open-source soundtouch libraries](https://github.com/also/soundtouch-js) years ago and I used it to create [this plunker.](https://plnkr.co/edit/ROQyEtPgQGfgZdluRzQy).

I'm no expert on the Web Audio API and cannot figure out how to integrate it smoothly with your fantastic library. This creates a script processing node and I figured you could just connect it in to your context. Obviously it's more complicated than that; I don't know how to make it work with play and pause OR getting things to line up with different playback rates.

Does this look promising?

**Comment by jsphweid** [comment]
@katspaugh Sorry, I can't stop playing around with this.

First off, I made somewhat of a working version with time-stretch preserve pitch: [current website](https://josephweidinger.com/browserTranscribe) & [GitHub repo](https://github.com/jsphweid/browserTranscribe).

But just to summarize, changes to WebAudio js file -> https://www.diffnow.com/?report=w2vuc 
It's kind of a mess, I know. But it works mostly well for my purposes. Assumes 2 channel audio (although I just fixed that on my site).

To be honest, I wonder if a lot of this is going to change when they stop supporting script processor nodes (https://webaudio.github.io/web-audio-api/#idl-def-ScriptProcessorNode) in favor of AudioWorklet nodes.

**Comment by katspaugh** [comment]
@jsphweid if you are willing to finish this (and create a PR), I know @jeremylach is ready to sponsor this feature.

**Comment by jeremylach** [comment]
Yes, a bounty can be arranged for this feature.

**Comment by jsphweid** [comment]
Ok, well I finally just finished [my web-app](https://www.josephweidinger.com/browserTranscribe/) that uses it but it looks really messy. I guess the best way to approach this would be to package up the SoundTouch files into a singular object and redo my changes as to not touch WaveSurfer.js, and extend it like a plugin or something.

Honestly, I'm not that great at this sort of thing but I'll give it a try soon.

**Comment by jeremylach** [comment]
@jsphweid Have you done any more work on this?

**Comment by jsphweid** [comment]
@jeremylach I'm afraid not. I'm too green in DSP to code this at the moment. Don't count on me :/

**Comment by kaicarver** [comment]
I also want to change rate while preserving pitch, and FYI I just found a way to do it with wavesurfer.js: just specify this initialization option in the `create() ` call: `backend: 'MediaElement'`.

I can't be the only person stymied and baffled by the differences between the HTML `<audio>` element and the Web Audio API. By default, wavesurfer uses WebAudio as a backend, but if I understand correctly this option tells it to use the `<audio>` element instead.

With my luck this will break or make impossible other wavesurfer features that work fine only with Web Audio, so I foresee tears in the future... but for the time being this looks like an ok kludgy workaround way to change rate while preserving pitch.

**Comment by jeremylach** [comment]
This does work @kaicarver however MediaElement does cause other bugs I've noticed.

**Comment by fxmarty** [comment]
Using MediaElement as a backend seems okay, however I don't really know the difference with WebAudio, but it fells globally not as smooth.

**Comment by sundayz** [comment]
MediaElement backend uses https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement which has the "preservesPitch" option. Changing tempo in WebAudio works differently

**Comment by ClemSK** [comment]
> I also want to change rate while preserving pitch, and FYI I just found a way to do it with wavesurfer.js: just specify this initialization option in the `create() ` call: `backend: 'MediaElement'`.
> 
> I can't be the only person stymied and baffled by the differences between the HTML `<audio>` element and the Web Audio API. By default, wavesurfer uses `WebAudio` as a backend, but if I understand correctly this option tells it to use the `<audio>` element instead.
> 
> With my luck this will break or make impossible other wavesurfer features that work fine only with Web Audio, so I foresee tears in the future... but for the time being this looks like an ok kludgy workaround way to change rate while preserving pitch.

I've used the MediaElement and it solves both the pitch-shifting and playback 'jumping' when toggling playback rate. Overall the HTML5 element is superior to the WabAudio API for delivering functionality. FYI as mentioned in another issue - I recommend setting partialRender: False as the MediaElement didn't render the wave if set to True in my case.

**Comment by ruchaheda** [comment]
Thank you for this discussion!

I am using WebAudio for my application with the wavesurfer API because, if I use MediaElement, the regions that I've created sometimes shift on load onto the track. For example, I have a ~40min track, and I find that there are certain regions that are consistently off by ~15seconds from when I initially created them. I confirmed that my file has a constant bit rate (after reading [this thread](https://github.com/katspaugh/wavesurfer.js/issues/2890)). My understanding is that WebAudio is better to use for more granular interactions on the audio file. 

Is there any way that I could continue to use WebAudio and also change the playback rate while preserving pitch?
